### ISSUE TITLE
Defind deleting command/operation what should do if selection range crosses editing host boundaries

### ISSUE THREAD

**Issue by masayuki-nakano** [w3c/editing#283]
I tested on Firefox and Chrome, Selection API add a range which crosses editing host boundaries in some cases. E.g.,
* `<div contenteditable><p>ab[c<span contenteditable="false">no]n-editable</span>def</p></div>`
* `<div contenteditable><p>a[bc<span contenteditable="false">non-editable<span contenteditable>de]f</span></span></p></div>`

If you `Backspace`, Firefox does nothing but fire `beforeinput` event, but Chrome deletes editable content **until** meeting first non-editable content.

I think that Firefox's behavior is better than Chrome for defining as standards and simpler. E.g., "If nearest inclusive ancestor editing host of start container and end container of a range is not same, nothing should be deleted, but dispatch `beforeinput` event whose target ranges is an empty array and whose input type is a value corresponding to the command (i.e., give the hints of delete amount and delete direction to web apps instead of using `deleteContent`).

I'll add tentative WPT about this issue, but it just checks whether the deleted range and target ranges are matched.

**Comment by masayuki-nakano** [comment]
If such range couldn't be in `Selection` with limitation of Selection API, it'd be better. But selecting all content may contain another editing host. Therefore, I think that this should be defined by editing API.

**Comment by masayuki-nakano** [comment]
> If such range couldn't be in `Selection` with limitation of Selection API, it'd be better. But selecting all content may contain another editing host. Therefore, I think that this should be defined by editing API.

Ah, this is not the case of this issue because range selecting all things may contain editing host boundaries, but does not cross them.

**Comment by johanneswilm** [comment]
> I think that Firefox's behavior is better than Chrome for defining as standards and simpler.

How about this situation:

A.  `<div contenteditable><p>ab[c<span contenteditable="false">non-editable]</span>def</p></div>`

vs.

B.  `<div contenteditable><p>ab[c<span contenteditable="false">non-editable</span>]def</p></div>`

?

If I understood you correctly, in situation A, nothing should be deleted, whereas in situation B, the entire selection should be deleted, correct?

I don't have a strong opinion, but I think one disadvantage of doing it the way Firefox does it is that a very small difference in the selection leads to a big difference in browser behavior. The difference in selection could be so small that the user doesn't even notice and therefore gets confused about why the selected content is not deleting.

**Comment by masayuki-nakano** [comment]
> 
> 
> > I think that Firefox's behavior is better than Chrome for defining as standards and simpler.
> 
> How about this situation:
> 
> A. `<div contenteditable><p>ab[c<span contenteditable="false">non-editable]</span>def</p></div>`
> 
> vs.
> 
> B. `<div contenteditable><p>ab[c<span contenteditable="false">non-editable</span>]def</p></div>`
> 
> ?
> 
> If I understood you correctly, in situation A, nothing should be deleted, whereas in situation B, the entire selection should be deleted, correct?

Correct. But I'm open to check whether the point is next to editable point.

> I don't have a strong opinion, but I think one disadvantage of doing it the way Firefox does it is that a very small difference in the selection leads to a big difference in browser behavior. The difference in selection could be so small that the user doesn't even notice and therefore gets confused about why the selected content is not deleting.

This shouldn't occur in these days since as you know, `contenteditable="false"` handling of browsers is different. And the cost to check whether the start and end of the range is next to editable content in same editing host can be ignored since it shouldn't be a hot path in real use cases. So, I'm wondering that this can be standardized.
